File: android/src/com/mygdx/game/voxel/FaceParts.java
package com.mygdx.game.voxel;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Pixmap;
import com.mygdx.game.voxel.reader.BinaryReader;
import com.mygdx.game.voxel.reader.VoxReader;
import com.mygdx.game.voxel.reader.VoxelData;

import java.security.MessageDigest;
import java.util.ArrayList;

/**
 * Created by daely on 11/28/2016.
 */
//TODO add three possible rotations
//TODO generate face background programmatically
    //TODO separate hair and hat block: each block that can have variable color should be independent
    //TODO single block depth for faces for saving memory: in fondo sono delle maschere!!
    //TODO put fragment in framelayout below another view
    //BLOCKS: eye block (with color)
    //??guancette rosse block
    //face shape block (with color)
    //bocca (with color)
    //ciglia (with color)
    //tatuaggio in faccia (with color)
    //capelli (uomo/donna
    //cappelli trecce
    //capelli basette
    //vestito (uomo donna) e colore
    //mano che tiene cose
    //ali bloc
    //cravatta block (con colore)
    //orecchini block (con colore)
    //occhiali block
    //animaletto block
    //cappello block

public class FaceParts {
    String basePath;
    ArrayList<String> partNames;
    ArrayList<Integer> partCounts;
    MersenneTwisterFast rand;
    int idDigestMode;
    public FaceParts(String basePath) {
        this.basePath=basePath;
        rand = null;
        idDigestMode=0;
        partCounts=new ArrayList<>();
        partNames = new ArrayList<>();
    }

    /**
     * search for file names from 1 to count
     * @param name
     * @param count
     */
    public void addPart(String name,int count) {
        partNames.add(name);
        partCounts.add(count);
    }
    public int nParts() { return  partCounts.size();}
    public VoxelData[] getFaceParts(String id) {
        int[] selectedParts = getFacePartIdxFromId(id);
        if (selectedParts == null) return null;
        int nparts=nParts();
        VoxelData[] res=new VoxelData[nparts];
        for (int i = 0; i < nparts; i++) {
            String pn=partNames.get(i);
            int[] defaultRGBAVoxPalette= getRGBAVoxelPalette(basePath,pn+"_palette.png");
            if(defaultRGBAVoxPalette==null) return null;
            VoxelData vox = getVoxelData(basePath, defaultRGBAVoxPalette,pn+selectedParts[i]+".vox");
            if(vox==null) return null;
            res[i]=vox;
        }
        return res;
    }

    private int[] getFacePartIdxFromId(String id) {
        byte[] digest=null;
        switch (idDigestMode) {
            case 0:
                digest= digestVerbatim(id);
                break;
            case 1:
                digest= digestMD5(id);
                break;
            case 2:
                digest= digestSHA256(id);
                break;
        }
        if(digest==null) return null;
        int[] seed=byteToInt(digest);
        MersenneTwisterFast mtf= new MersenneTwisterFast(seed);
        int nparts_=nParts();
        int[] selectedParts=new int[nparts_];
        for (int i = 0; i < nparts_; i++) {
            selectedParts[i]=mtf.nextInt(partCounts.get(i));
        }
        return selectedParts;
    }

    private int[] byteToInt(byte[] bytes) {
        int len=bytes.length;
       int[] ints=new int[len];
        for (int i = 0; i < len; i++) {
            ints[i]=bytes[i];
        }
        return ints;
    }

    private byte[] digestVerbatim(String id) {
        try {
            return id.getBytes("UTF-8");
        } catch (Exception e) {
            return null;
        }
    }
    private byte[] digestMD5(String id) {
        byte[] digest=null;
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            digest = md.digest(id.getBytes("UTF-8"));
        } catch (Exception e) {
            return null;
        }
        return digest;
    }
    private byte[] digestSHA256(String id) {
        byte[] digest=null;
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            digest = md.digest(id.getBytes("UTF-8"));
        } catch (Exception e) {
            return null;
        }
        return digest;
    }
    private VoxelData getVoxelData(String voxpath, int[] defaultRGBAVoxPalette, String voxname) {
        FileHandle vxf1= Gdx.files.internal(voxpath+ voxname);
        return VoxReader.fromMagica(new BinaryReader(vxf1),defaultRGBAVoxPalette);
    }
    private int[] getRGBAVoxelPalette(String voxpath, String voxpalettename) {
        FileHandle vxf1= Gdx.files.internal(voxpath+ voxpalettename);
        Pixmap pixmap= new Pixmap(vxf1);
        int[] res= new int[256];
        for (int i = 0; i < 256; i++) {
            res[i]=pixmap.getPixel(i,0);
        }
        return res;
    }
}

File: android/src/com/mygdx/game/voxel/FaceParts.kt
package com.mygdx.game.voxel

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.graphics.Pixmap
import com.mygdx.game.voxel.reader.BinaryReader
import com.mygdx.game.voxel.reader.VoxReader
import com.mygdx.game.voxel.reader.VoxelData

import java.security.MessageDigest
import java.util.ArrayList

/**
 * Created by daely on 11/28/2016.
 */
//TODO add three possible rotations
//TODO generate face background programmatically
//TODO separate hair and hat block: each block that can have variable color should be independent
//TODO single block depth for faces for saving memory: in fondo sono delle maschere!!
//TODO put fragment in framelayout below another view
//BLOCKS: eye block (with color)
//??guancette rosse block
//face shape block (with color)
//bocca (with color)
//ciglia (with color)
//tatuaggio in faccia (with color)
//capelli (uomo/donna
//cappelli trecce
//capelli basette
//vestito (uomo donna) e colore
//mano che tiene cose
//ali bloc
//cravatta block (con colore)
//orecchini block (con colore)
//occhiali block
//animaletto block
//cappello block

class FaceParts(internal var basePath: String) {
    internal var partNames: ArrayList<String>
    internal var partCounts: ArrayList<Int>
    internal var rand: MersenneTwisterFast? = null
    internal var idDigestMode: Int = 0

    init {
        rand = null
        idDigestMode = 1 //MD5
        partCounts = ArrayList<Int>()
        partNames = ArrayList<String>()
    }

    /**
     * search for file names from 1 to count
     * @param name
     * *
     * @param count
     */
    fun addPart(name: String, count: Int) {
        partNames.add(name)
        partCounts.add(count)
    }

    fun nParts(): Int {
        return partCounts.size
    }

    fun getFaceParts(id: String): Array<VoxelData>? {
        val selectedParts = getFacePartIdxFromId(id) ?: return null
        val nparts = nParts()
        val res = Array<VoxelData>(nparts) { ipart ->
            val pn = partNames[ipart]
            val defaultRGBAVoxPalette = getRGBAVoxelPalette(basePath, pn + "_palette.png") ?: return null
            val vox = getVoxelData(basePath, defaultRGBAVoxPalette, pn + selectedParts[ipart] + ".vox") ?: return null
            vox
        }
        return res
    }

    private fun getFacePartIdxFromId(id: String): IntArray? {
        var digest: ByteArray? = null
        when (idDigestMode) {
            0 -> digest = digestVerbatim(id)
            1 -> digest = digestMD5(id)
            2 -> digest = digestSHA256(id)
        }
        if (digest == null) return null
        val seed = byteToInt(digest)
        val mtf = MersenneTwisterFast(seed)
        val nparts_ = nParts()
        val selectedParts = IntArray(nparts_)
        for (i in 0..nparts_ - 1) {
            selectedParts[i] = mtf.nextInt(partCounts[i])
        }
        return selectedParts
    }

    private fun byteToInt(bytes: ByteArray): IntArray {
        val len = bytes.size
        val ints = IntArray(len)
        for (i in 0..len - 1) {
            ints[i] = bytes[i].toInt()
        }
        return ints
    }

    private fun digestVerbatim(id: String): ByteArray? =
            try {
                id.toByteArray(charset("UTF-8"))
            } catch (e: Exception) {
                null
            }

    private fun digestMD5(id: String): ByteArray? =
            try {
                val md = MessageDigest.getInstance("MD5")
                md.digest(id.toByteArray(charset("UTF-8")))
            } catch (e: Exception) {
                null
            }

    private fun digestSHA256(id: String): ByteArray? =
            try {
                val md = MessageDigest.getInstance("SHA-256")
                md.digest(id.toByteArray(charset("UTF-8")))
            } catch (e: Exception) {
                null
            }

    private fun getVoxelData(voxpath: String, defaultRGBAVoxPalette: IntArray, voxname: String): VoxelData? {
        val vxf1 = Gdx.files.internal(voxpath + voxname)
        return VoxReader.fromMagica(BinaryReader(vxf1), defaultRGBAVoxPalette)
    }

    private fun getRGBAVoxelPalette(voxpath: String, voxpalettename: String): IntArray? {
        val vxf1 = Gdx.files.internal(voxpath + voxpalettename)
        val pixmap = Pixmap(vxf1)
        val res = IntArray(256)
        for (i in 0..255) {
            res[i] = pixmap.getPixel(i, 0)
        }
        return res
    }
}
